operator/duties/fetcher.go

```diff
@@ -0,0 +1,180 @@
+			duties = raw.(dutyCacheEntry).Duties
+		}
+	}
+
+	if len(duties) > 0 {
+		logger.Debug("found duties for slot",
+			zap.Int("count", len(duties)), zap.Any("duties", duties))
+	} else {
+		logger.Debug("didn't found duties for slot")
+	}
+
+	return duties, nil
+}
+
+// updateDutiesFromBeacon will be called once in an epoch to update the cache with all the epoch's slots
+func (df *dutyFetcher) updateDutiesFromBeacon(slot uint64) error {
+	attesterDuties, err := df.fetchAttesterDuties(slot)
+	if err != nil {
+		return errors.Wrap(err, "failed to get attest duties")
+	}
+	df.logger.Debug("got duties", zap.Int("count", len(attesterDuties)), zap.Any("attesterDuties", attesterDuties))
+	if err := df.processFetchedDuties(attesterDuties); err != nil {
+		return errors.Wrap(err, "failed to process fetched duties")
+	}
+	return nil
+}
+
+// fetchAttesterDuties fetches duties for the epoch of the given slot
+func (df *dutyFetcher) fetchAttesterDuties(slot uint64) ([]*eth2apiv1.AttesterDuty, error) {
+	if indices := df.indicesFetcher.GetValidatorsIndices(); len(indices) > 0 {
+		df.logger.Debug("got indices for existing validators",
+			zap.Int("count", len(indices)), zap.Any("indices", indices))
+		esEpoch := df.ethNetwork.EstimatedEpochAtSlot(slot)
+		epoch := spec.Epoch(esEpoch)
+		return df.beaconClient.GetDuties(epoch, indices)
+	}
+	df.logger.Debug("got no indices, duties won't be fetched")
+	return []*eth2apiv1.AttesterDuty{}, nil
+}
+
+// processFetchedDuties loop over fetched duties and process them
+func (df *dutyFetcher) processFetchedDuties(attesterDuties []*eth2apiv1.AttesterDuty) error {
+	var subscriptions []*eth2apiv1.BeaconCommitteeSubscription
+	entries := map[spec.Slot]dutyCacheEntry{}
+	for _, ad := range attesterDuties {
+		duty := convertAttesterDuty(ad)
+		df.createEntry(entries, duty)
+		subscriptions = append(subscriptions, toSubscription(duty))
+	}
+	df.populateCache(entries)
+	if err := df.beaconClient.SubscribeToCommitteeSubnet(subscriptions); err != nil {
+		df.logger.Warn("failed to subscribe committee to subnet", zap.Error(err))
+	}
+	return nil
+}
+
+func (df *dutyFetcher) createEntry(entries map[spec.Slot]dutyCacheEntry, duty *beacon.Duty) {
+	entry, slotExist := entries[duty.Slot]
+	if !slotExist {
+		entry = dutyCacheEntry{[]beacon.Duty{}}
+	}
+	entry.Duties = append(entry.Duties, *duty)
+	entries[duty.Slot] = entry
+}
+
+// populateCache takes a map of entries and updates the cache
+func (df *dutyFetcher) populateCache(entriesToAdd map[spec.Slot]dutyCacheEntry) {
+	for s, e := range entriesToAdd {
+		slot := uint64(s)
+		if raw, exist := df.cache.Get(getDutyCacheKey(slot)); exist {
+			existingEntry := raw.(dutyCacheEntry)
+			e.Duties = append(existingEntry.Duties, e.Duties...)
+		}
+		df.cache.SetDefault(getDutyCacheKey(slot), e)
+	}
+}
+
+func getDutyCacheKey(slot uint64) string {
+	return fmt.Sprintf("%d", slot)
+}
+
+func convertAttesterDuty(attesterDuty *eth2apiv1.AttesterDuty) *beacon.Duty {
```

Comment:
@amir-blox why are we using 2 different types of structs for duties?

Reply:
I was following https://github.com/bloxapp/ssv/blob/main/beacon/goclient/goclient.go#L87 
This could be handled when we introduce multi-duty support